Source/Common/SDecode.cpp

```diff
@@ -0,0 +1,52 @@
+

```

Comment:
MothershipDummy is, as you say, not interacting with the HostLink, but in addition, it's configured to respond to OnReply messages (which are responses to NameServer requests). 

This is particularly useful for testing because it means you don't need a querying external process (which would have to be handled through the nonexistent UserIO process), you can just issue queries from MothershipDummy and then get responses back from the NameServer. The ConfigState command (Q::NAME,Q::CFG,Q::STATE) in the message key) triggers MothershipDummy to run some queries. 

Meanwhile, a regular Mothership should definitely not respond like MothershipDummy does to OnReply messages - it should just invoke the SBase class' default handler for that (warn and continue), and it MUST not respond to ConfigState messages (because the  state of a Mothership explicitly depends on actions that interact with the HostLink - thus a command from outside to change its state should do nothing), so deriving from MothershipDummy is not appropriate. 

You are correct that it's there mostly for testing - to emulate within reason a 'Mothership' without actually being a HostLink as well. I suspect though that when we come to test UserIO it will be valuable in any case. MothershipDummy also shows how a lot of the system was tested - by deploying tasks to it and having it output to a file the results it received after running a series of queries (we have to test the query functionality!)

Reply:
After looking through this carefully I am coming to the conclusion that there is no easy way to do this. The issues:

You can't use the same FnMapx because the derived classes need different dispatch routines for subsets of the commands.

The 'obvious' way to modify Decode.cpp, by testing on a `dynamic_cast<SBase*>(this) != 0`, runs into the issue that you then need to include the SBase header in all the processes' .h files. And the line `return (this->*(*SBase::FnMapx[cIdx])[pPkt->Key()])(pPkt,cIdx);` I think might still fall over or fail to compile, if the processes themselves aren't actually derived from SBase.

The 'other' way of modifying it, using an #ifdef __SBASE_H__ around the crucial section, is an ugly compiler-directive bodge, which I really don't like. I personally think that's MUCH worse than using 2 different decode includes. At least using 2 different files makes it clear what's going to happen in the control logic for any given class.

Decode has always been an exceptionally tricky piece of code, fundamentally it's attempting to achieve something C++ really, really doesn't want to let you do. However, it's so deeply embedded into the functionality of the Orchestrator that at this point, trying to do something entirely different would be a serious effort.

Overall, attempting some sort of alternative route here looks to be going down the line of a very significant rewrite, which I think is not in scope